File: 0001-0099/0035. Search Insert Position/0035. Search Insert Position.py
from typing import List


class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        lo, hi = 0, len(nums) - 1

        while lo <= hi:
            mid = lo + (hi - lo) // 2
            if nums[mid] < target:
                lo = mid + 1
            elif nums[mid] > target:
                hi = mid - 1
            else:
                return mid
        # 这里的 lo > hi => lo >= hi + 1
        # 插入的index位置只能是 [lo,hi+1]
        # 即  lo >= hi + 1  and  lo <= hi + 1
        # 即 lo == hi + 1
        # 插入的位置只能为lo
        return lo

File: 0001-0099/0035. Search Insert Position/0035. Search Insert Position.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var searchInsert = function (nums, target) {
    let [low, high] = [0, nums.length - 1];
    while (low <= high) {
        const mid = low + ((high - low) >> 1);
        if (nums[mid] === target) return mid;
        else if (nums[mid] > target) high = mid - 1;
        else low = mid + 1;
    }
    return low;
};
// 最后结束前的位置 
// low + 1 === high  mid === low
//     mid > target high = mid - 1 === low - 1  跳出循环 low > target return low 
//     mid < target low = mid + 1 === low + 1  low === high 进行下次循环
// low === high mid === high === low
//     mid > target => low > target  return low 
//     mid < target => low < target 应该插入到 low + 1 去 也就是 mid + 1